File: js/shitjsvm/js/classLoader.js
import { JClass } from "./class/JClass"
import { Constant } from "./class/Constant"
import { Field } from "./class/Field"
import { Attribute } from "./class/Attribute"
import { Method } from "./class/Method"
import {
    CONSTANT_Class,
    CONSTANT_Double, CONSTANT_Fieldref,
    CONSTANT_Float,
    CONSTANT_Integer, CONSTANT_InterfaceMethodref,
    CONSTANT_Long, CONSTANT_Methodref, CONSTANT_NameAndType, CONSTANT_String,
    CONSTANT_Utf8
} from "./const/constNames"

let index = 0

function resolve(constants, i) {
    const c = constants[i]
    if (c == null) throw Error("Invalid constant index: " + i)
    if (c.tag === CONSTANT_Utf8) return c.value.toString()
    if (c.tag === CONSTANT_Class || c.tag === CONSTANT_String) return resolve(constants, c.value)
        throw Error("Invalid constant tag: 0x0" + c.tag.toString(16).toUpperCase())
}

function signature(data) {
    const signature = data.getUint32(index, false)
    index += 4
    if (signature !== 0xCAFEBABE) throw Error("Invalid signature")

    return signature
}

function minor(data) {
    const minor = data.getUint16(index, false)
    index += 2

    return minor
}

function major(data) {
    const major = data.getUint16(index, false)
    index += 2

    return major
}

function constants(data) {
    const constantCount = data.getUint16(index, false)
    index += 2
    const constants = new Array(constantCount)
    for (let i = 1; i < constantCount; i++) {
        const tag = data.getUint8(index)
        index += 1
        switch (tag) {
            case CONSTANT_Utf8:
                const length = data.getUint16(index, false)
                index += 2
                const stringData = new Uint8Array(length)
                for (let j = 0; j < length; j++) {
                    stringData[j] = data.getUint8(index + j)
                }
                index += length

                const decoder = new TextDecoder()
                const stringValue = decoder.decode(stringData)

                constants[i] = new Constant(tag, stringValue)
                break
            case CONSTANT_Integer:
                constants[i] = new Constant(tag, data.getInt32(index, false))
                index += 4
                break
            case CONSTANT_Float:
                constants[i] = new Constant(tag, data.getFloat32(index, false))
                index += 4
                break
            case CONSTANT_Long:
                constants[i] = new Constant(tag, data.getBigInt64(index, false))
                index += 8
                i++
                if (i < constantCount)
                    constants[i] = null
                break
            case CONSTANT_Double:
                constants[i] = new Constant(tag, data.getFloat64(index, false))
                index += 8
                i++
                if (i < constantCount)
                    constants[i] = null
                break
            case CONSTANT_Class: case CONSTANT_String:
                constants[i] = new Constant(tag, data.getUint16(index, false))
                index += 2
                break
            case CONSTANT_Fieldref: case CONSTANT_Methodref: case CONSTANT_InterfaceMethodref:
                const classIndex = data.getUint16(index, false)
                index += 2
                const nameAndTypeIndex = data.getUint16(index, false)
                index += 2
                constants[i] = new Constant(tag, [classIndex, nameAndTypeIndex])
                break
            case CONSTANT_NameAndType:
                const nameIndex = data.getUint16(index, false)
                index += 2
                const descriptorIndex = data.getUint16(index, false)
                index += 2
                constants[i] = new Constant(tag, [nameIndex, descriptorIndex])
                break
            default:
                throw Error("Unknown constant: " + tag.toString(16).toUpperCase())
        }
    }

    return constants
}

function flags(data) {
    const flags = data.getInt16(index)
    index += 2

    return flags
}

function interfaces(data, constants) {
    const interfaceCount = data.getUint16(index, false)
    index += 2
    const interfaces = []
    for (let i = 0; i < interfaceCount; i++) {
        interfaces.push(resolve(constants, data.getUint16(index, false)))
        index += 2
    }

    return interfaces
}

function fields(data, constants) {
    const fieldCount = data.getUint16(index, false)
    index += 2
    const fields = []
    for (let i = 0; i < fieldCount; i++) {
        const flags = data.getUint16(index, false)
        index += 2
        const name = resolve(constants, data.getUint16(index, false))
        index += 2
        const descriptor = resolve(constants, data.getUint16(index, false))
        index += 2
        const attribs = attributes(data, constants)

        fields.push(new Field(flags, name, descriptor, attribs))
    }

    return fields
}

function attributes(data, constants) {
    const attributeCount = data.getUint16(index, false)
    index += 2
    const attributes = []
    for (let i = 0; i < attributeCount; i++) {
        const name = resolve(constants, data.getUint16(index, false))
        index += 2
        const length = data.getInt32(index, false)
        index += 4
        const value = new Uint8Array(length)
        for (let j = 0; j < length; j++) {
            value[j] = data.getUint8(index + j)
        }
        index += length

        attributes.push(new Attribute(
            name,
            value
        ))
    }

    return attributes
}

function methods(data, constants) {
    const methodCount = data.getUint16(index, false)
    index += 2
    const methods = []
    for (let i = 0; i < methodCount; i++) {
        const flags = data.getUint16(index, false)
        index += 2
        const name = resolve(constants, data.getUint16(index, false))
        index += 2
        const descriptor = resolve(constants, data.getUint16(index, false))
        index += 2
        const attribs = attributes(data, constants)

        methods.push(new Method(flags, name, descriptor, attribs))
    }

    return methods
}

export function decode(data) {
    index = 0

    const vClass = new JClass()

    vClass.signature = signature(data)
    vClass.minor = minor(data)
    vClass.major = major(data)
    vClass.constants = constants(data)
    vClass.flags = flags(data)
    vClass.name = resolve(vClass.constants, data.getUint16(index, false))
    index += 2
    vClass.superClass = resolve(vClass.constants, data.getUint16(index, false))
    index += 2
    vClass.interfaces = interfaces(data, vClass.constants)
    vClass.fields = fields(data, vClass.constants)
    vClass.methods = methods(data, vClass.constants)
    vClass.attributes = attributes(data, vClass.constants)

    return vClass
}

File: java/src/main/java/dev/blackoutburst/shitvm/ClassLoader.java
package dev.blackoutburst.shitvm;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ClassLoader {

    private String resolve(List<Constant> constants, int index) {
        Constant c = constants.get(index);
        if (c == null) return "unknown";
        if (c.tag == 0x01) {
            return (String) c.value;
        } else {
            return "unknown";
        }
    }


    private List<Constant> constants(ByteBuffer data) throws Exception {
        short constantCount = data.getShort();
        List<Constant> constants = new ArrayList<>(Collections.nCopies(constantCount, null));
        for (int i = 1; i < constantCount; i++) {
            byte tag = data.get();
            switch (tag) {
                case 0x01:
                    byte[] stringData = new byte[data.getShort() & 0xFFFF];
                    data.get(stringData);
                    constants.set(i,new Constant(tag, new String(stringData)));
                    break;
                case 0x03:
                    constants.set(i,new Constant(tag, data.getInt()));
                    break;
                case 0x04:
                    constants.set(i,new Constant(tag, data.getFloat()));
                    break;
                case 0x05:
                    constants.set(i,new Constant(tag, data.getLong()));
                    i++;
                    if (i < constantCount) {
                        constants.set(i, null);
                    }
                    break;
                case 0x06:
                    constants.set(i,new Constant(tag, data.getDouble()));
                    i++;
                    if (i < constantCount) {
                        constants.set(i, null);
                    }
                    break;
                case 0x07, 0x08:
                    constants.set(i,new Constant(tag, data.getShort() & 0xFFFF));
                    break;
                case 0x09, 0x0A, 0x0B:
                    int classIndex = data.getShort() & 0xFFFF;
                    int nameAndTypeIndex = data.getShort() & 0xFFFF;
                    constants.set(i,new Constant(tag, new int[]{classIndex, nameAndTypeIndex}));
                    break;
                case 0x0C:
                    int nameIndex = data.getShort() & 0xFFFF;
                    int descriptorIndex = data.getShort() & 0xFFFF;
                    constants.set(i,new Constant(tag, new int[]{nameIndex, descriptorIndex}));
                    break;
                default:
                    throw new Exception("Unknown constant pool tag: " + tag);
            }
        }

        return constants;
    }

    private List<String> interfaces(ByteBuffer data, List<Constant> constants) {
        List<String> interfaces = new ArrayList<>();
        short interfaceCount = data.getShort();
        for (int i = 0; i < interfaceCount; i++) {
            interfaces.add(resolve(constants, data.getShort()));
        }

        return interfaces;
    }

    private List<Field> fields(ByteBuffer data, List<Constant> constants) {
        List<Field> fields = new ArrayList<>();
        short fieldCount = data.getShort();
        for (int i = 0; i < fieldCount; i++) {
            fields.add(new Field(
                    data.getShort(),
                    resolve(constants, data.getShort()),
                    resolve(constants, data.getShort()),
                    attributes(data, constants)
            ));
        }

        return fields;
    }

    private List<Attribute> attributes(ByteBuffer data, List<Constant> constants) {
        short attributeCount = data.getShort();
        List<Attribute> attribs = new ArrayList<>();
        for (int i = 0; i < attributeCount; i++) {
            String name = resolve(constants, data.getShort());
            byte[] value = new byte[data.getInt()];
            data.get(value);

            attribs.add(new Attribute(
                    name,
                    value
            ));
        }

        return attribs;
    }

    private List<Method> methods(ByteBuffer data, List<Constant> constants) {
        List<Method> methods = new ArrayList<>();
        short methodCount = data.getShort();
        for (int i = 0; i < methodCount; i++) {
            short accessFlags = data.getShort();
            String name = resolve(constants, data.getShort());
            String descriptor = resolve(constants, data.getShort());
            List<Attribute> attributes = attributes(data, constants);

            methods.add(new Method(accessFlags, name, descriptor, attributes));
        }
        return methods;
    }


    private Frame createFrame(VClass vClass, String method, Object... args) {
        for (Method m : vClass.methods) {
            if (!m.name.equals(method)) continue;
            for (Attribute a : m.attributes) {
                if (a.name.equals("Code") && a.data.length > 8) {
                    Frame frame = new Frame(
                            vClass,
                            0,
                            Arrays.copyOfRange(a.data, 8, a.data.length)
                    );
                    frame.locals.addAll(Arrays.asList(args));
                    return frame;
                }
            }
        }
        return null;
    }

    private Object execute(Frame frame) {
        while (true) {
            byte op = frame.code[frame.ip];
            System.out.println(op);
            switch (op) {
                case 26:
                    frame.stack.add(frame.locals.get(0));
                    break;
                case 27:
                    frame.stack.add(frame.locals.get(1));
                    break;
                case 96:
                    int a = (int) frame.stack.getLast();
                    frame.stack.removeLast();
                    int b = (int) frame.stack.getLast();
                    frame.stack.removeLast();
                    frame.stack.add(a + b);
                    break;
                case 100:
                    int a2 = (int) frame.stack.getLast();
                    frame.stack.removeLast();
                    int b2 = (int) frame.stack.getLast();
                    frame.stack.removeLast();
                    frame.stack.add(a2 - b2);
                    break;
                case -84:
                    Object v = frame.stack.getLast();
                    frame.stack.removeLast();
                    return v;
            }
            frame.ip++;
        }
    }

    public VClass decode(ByteBuffer data) throws Exception {
        VClass vClass = new VClass();
        vClass.signature = data.getInt();
        if (vClass.signature != 0xCAFEBABE) throw new Exception();
        vClass.minor = data.getShort();
        vClass.major = data.getShort();
        vClass.constants = constants(data);
        vClass.flags = data.getShort();
        vClass.name = resolve(vClass.constants, data.getShort());
        vClass.superClass = resolve(vClass.constants, data.getShort());
        vClass.interfaces = interfaces(data, vClass.constants);
        vClass.fields = fields(data, vClass.constants);
        vClass.methods = methods(data, vClass.constants);
        vClass.attributes = attributes(data, vClass.constants);

        System.out.println("Signature: " + vClass.signature);
        System.out.println("Minor: " + vClass.minor);
        System.out.println("Major: " + vClass.major);
        System.out.println("Name: " + vClass.name);
        System.out.println("Super: " + vClass.superClass);
        System.out.println("Flags: " + vClass.flags);
        System.out.print("Fields: ");
        for (Field m : vClass.fields) {
            System.out.print(m.name + ", ");
        }
        System.out.println();
        System.out.println("Methods: ");
        for (Method m : vClass.methods) {
            System.out.println(m.name);
            System.out.println("Methods Attributes: ");
            for (Attribute a : m.attributes) {
                System.out.print(a.name + ", ");
            }
        }
        System.out.println();
        System.out.print("Class Attributes: ");
        for (Attribute m : vClass.attributes) {
            System.out.print(m.name + ", ");
        }
        System.out.println();

        Frame frame = createFrame(vClass, "sub", 30, 24);
        int result = (int) execute(frame);
        System.out.println("Result:" + result);
        return vClass;
    }

    public ByteBuffer loadClass(String className) throws IOException {
        return ByteBuffer.wrap(Files.readAllBytes(Path.of(className)));
    }
}
